### ISSUE TITLE
Threshold useless?

### ISSUE THREAD

**Issue by AzazelHD** [Akhilesh64/Image-Segmentation-U-2-Net#1]
What is threshold for?

I am not seeing any differences by modifying it

**Comment by Akhilesh64** [comment]
@AzazelHD Depends on the frame you are trying to threshold. If the resultant frame is distorted threshold helps in better binarization. You can skip out the thresholding if you dont want to use it.

**Comment by AzazelHD** [comment]
Mmmm
I am following this code

```
# For thresholding
for i in range(len(preds)):
    shape = preds[i, :, :].shape
    frame = binarize(preds[1, :, :], threshold=0.5)
    frame = np.reshape(frame, (shape[0], shape[1]))
```

And changing threshold doesn't do anything. 

Also, i am getting 7 mask layers. But I don't know which one offers better results. Or if I have to combine them all or whatever

**Comment by Akhilesh64** [comment]
Are you saving the frames after thresholding ?? This code snippet just thresholds the frames, you have to save them too. My overall code thresholds the frames but saves the original non-thresholded frames. Change the code to save the thresholded frames.

About the mask layers you recieve the first mask layer(say im0.png) is the combination of the remaining 6 mask layers(say im1.png to im6.png) predicted by the model.
Usually the the foremost frame (im1.png in this case) would have the best results.

**Comment by AzazelHD** [comment]
Yeah. I am saving the 7 frames.
img0 for me is completely white tho.

And back again to thresholding... Could you explain better please? I will dont see any changes if it is 0.1, 0.5 or 0.9

**Comment by Akhilesh64** [comment]
I need your minimum working code and the image you are using with the model. Thresholding is used to binarize the image better if your resultant frames contains only black or only white pixels thresholding wont do anything. Refer to this [link](https://opencv-python-tutroals.readthedocs.io/en/latest/py_tutorials/py_imgproc/py_thresholding/py_thresholding.html) for better understanding of thesholding.

**Comment by AzazelHD** [comment]
```
import cv2
import numpy as np
from keras.models import load_model
from sklearn.preprocessing import binarize

model = load_model('models/model_DUTS.h5', compile=False)

img = cv2.imread(r'images/spawn.png', cv2.IMREAD_COLOR)
img = cv2.resize(img, (256, 256))
img = np.expand_dims(img, axis=0)

preds = model.predict(img)
preds = np.squeeze(preds)

# For thresholding
for i in range(len(preds)):
    shape = preds[i, :, :].shape
    frame = binarize(preds[1, :, :], threshold=0.5)
    frame = np.reshape(frame, (shape[0], shape[1]))

# For saving all the frames
for i in range(len(preds)):
    img = np.expand_dims(np.array(preds[i, :, :]), axis=-1)
    img *= 255
    # img = np.where(img[:, :, 0] < 0.035, 0, 255).astype('uint8')
    cv2.imwrite(r'img' + str(i) + '.png', img)


src1 = cv2.imread('images/spawn.png', cv2.IMREAD_COLOR)
src1 = cv2.resize(src1, (256, 256))
src2 = cv2.imread('img1.png')
dst = cv2.bitwise_and(src1, src2)
cv2.imwrite('bitwise.jpg', dst)
```

Image are this ones: 
![spawn](https://user-images.githubusercontent.com/47080553/103543083-3a02e680-4e9e-11eb-86fe-89201647bc3e.png)
![spawn1](https://user-images.githubusercontent.com/47080553/103543086-3c654080-4e9e-11eb-9283-73f4d2ab3b87.png)

I think this last one is gonna be hard, but still, something is better than nothing

**Comment by Akhilesh64** [comment]
Try the following code and change the threshold value to see the change in resultant frames. You didn't apply what I said earlier :
> Are you saving the frames after thresholding ?? This code snippet just thresholds the frames, you have to save them too. My overall code thresholds the frames but saves the original non-thresholded frames. Change the code to save the thresholded frames.
> 
> About the mask layers you recieve the first mask layer(say im0.png) is the combination of the remaining 6 mask layers(say im1.png to im6.png) predicted by the model.
> Usually the the foremost frame (im1.png in this case) would have the best results.

```
import cv2
import numpy as np
from keras.models import load_model
from sklearn.preprocessing import binarize
from tensorflow.keras.preprocessing.image import array_to_img

model = load_model('models/model_DUTS.h5', compile=False)

img = cv2.imread('images/spawn.png', cv2.IMREAD_COLOR)
img = cv2.resize(img, (256, 256))
img = np.expand_dims(img, axis=0)

preds = model.predict(img)
preds = np.squeeze(preds)

# For thresholding
for i in range(len(preds)):
    shape = preds[i, :, :].shape
    frame = binarize(preds[1, :, :], threshold=0.5)
    frame = np.reshape(frame, (shape[0], shape[1]))
    frame = np.expand_dims(np.array(preds[i,:,:]),axis=-1)
    frame = array_to_img(frame)
    frame.save('img'+str(i)+'.png')
````
